// @talkjs/react-components/globalConstants.js
"use client";
import { createContext as n, useContext as o } from "react";
import { Schema as l } from "prosemirror-model";
class i extends Error {
}
function r(t, e) {
  if (!t) throw new i(e != null ? e : "Assertion failed");
}
const s = Object.assign(r, { is: function(t, e, c) {
  return r(e, c);
}, unreachable: function() {
  throw new Error("Unreachable code");
}, never: function(t) {
  throw new Error("Unreachable");
}, defined: function(t) {
  if (t == null) throw new i("Value is not defined");
} }), [g, x] = a(), [h, O] = a(), [C, D] = a(), [M, f] = a(), [j, b] = a(), m = n(void 0);
function v() {
  const t = o(m);
  return s(t !== void 0), t;
}
const u = n(void 0);
function w() {
  const t = o(u);
  return s(t !== void 0), t;
}
const [E, k] = a(), [P, A] = a();
function a() {
  const t = n(void 0);
  return [t, () => {
    const e = o(t);
    return s(e !== void 0), e;
  }];
}
const S = new l({ nodes: { doc: { content: "paragraph+" }, paragraph: { code: !0, content: "inline*", parseDOM: [{ tag: "p" }], toDOM: () => ["p", 0] }, text: { group: "inline" }, mention: { group: "inline", inline: !0, attrs: { id: {}, name: {} }, content: "text*", toDOM: (t) => ["span", { "data-mention-id": t.attrs.id, class: "t-mention" }, `@${t.attrs.name}`], parseDOM: [{ tag: "span[data-mention-id]", getAttrs: (t) => ({ id: t.dataset["mention-id"] }) }] }, customemoji: { group: "inline", inline: !0, attrs: { name: {}, url: {} }, content: "text*", selectable: !1, toDOM: (t) => ["img", { src: t.attrs.url, alt: t.attrs.name, "data-custom-emoji": t.attrs.name }], parseDOM: [{ tag: "img[data-custom-emoji]", getAttrs: (t) => {
  const e = t;
  return { name: e.dataset.customEmoji, url: e.src };
} }] } }, marks: { literal: { toDOM: () => ["span", { "data-literal": "" }], parseDOM: [{ tag: "span[data-literal]" }] }, bold: { toDOM: () => ["b"], parseDOM: [{ tag: "b" }] }, italic: { toDOM: () => ["i"], parseDOM: [{ tag: "i" }] }, strikethrough: { toDOM: () => ["del"], parseDOM: [{ tag: "del" }] }, emoji: { attrs: { emoji: {} }, toDOM: (t) => ["span", { "data-emoji": t.attrs.emoji }, t.attrs.emoji], parseDOM: [{ tag: "span[data-emoji]", getAttrs: (t) => ({ emoji: t.dataset.emoji }) }] } } });
export {
  j as ChatboxCallbacksContext,
  u as ControllerContext,
  h as EditorInternalsContext,
  E as EmojiPickerContext,
  C as PermissionContext,
  P as PopoverContext,
  M as SessionContext,
  m as StoreContext,
  g as ThemeContext,
  S as schema,
  b as useChatboxCallbacks,
  w as useController,
  O as useEditorInternals,
  k as useEmojiPicker,
  D as usePermission,
  A as usePopover,
  f as useSession,
  v as useStore,
  x as useTheme
};
